### ISSUE TITLE
Twitter Freezes and Crashes when GT is enabled

### ISSUE THREAD

**Issue by JesperDramsch** [Bl4Cc4t/GoodTwitter2#455]
**Please check the following before you continue**
- [x] Bug happens with the newest version of GT2 (see the [changelog](https://github.com/Bl4Cc4t/GoodTwitter2/blob/master/doc/changelog.md))
- [x] Checked [FAQ](https://github.com/Bl4Cc4t/GoodTwitter2#faq)
- [x] Checked if the issue has not been reported before

This is related to #454 but much more extensive, hence the new issue. (Sorry if that was wrong)

**Describe the bug**
Twitter breaks on any site I open when GT is enabled. The site is unresponsive to the point that I can't even open the console with F12 or click anything. It also affects the Twemex extension (however, disabling twemex doesn't solve it).

When I disable GT Twitter loads fine.

When I let it crash by itself eventually it gives the code "result_code_hung", but that hasn't helped debugging it so far.

**Screenshots**
![twitter-breaks](https://user-images.githubusercontent.com/2620316/183009463-3d45738e-1987-4013-b19f-7b23143ae5e5.gif)

**Your configuration**
 - Brave v1.42.86
 - Violentmonkey Version 2.13.0

**Comment by Mavoy** [comment]
I had similar issues on Chrome this morning, but interesingly enough, they disppeared after I updated to the latest version (from 103 to 104). I'm not sure about other browsers

**Comment by arturo182** [comment]
Have the same issue with latest Opera (89.0.4447.83). Downgrading GT2 to v0.0.39 doesn't fix it, it must be something that changed either in the Chromium engine or in Twitter itself.

**Comment by Mavoy** [comment]
And I spoke prematurely. It worked once after the update, but then it stopped. Hopefully it'll be fixed.

**Comment by jesusmgg** [comment]
Unusable on Firefox 103.0.1 currently.

**Comment by arturo182** [comment]
Fairly sure Firefox isn't Chromium-based, so it would mean it's a twitter.com change that caused this 🤔

**Comment by JesperDramsch** [comment]
> Fairly sure Firefox isn't Chromium-based, so it would mean it's a twitter.com change that caused this 🤔

Yes that's correct. Edge, Brave, and Chrome are Chromium. Firefox is its own thing.

**Comment by maxymax13** [comment]
currently unusable on FF 103.0.1, makes the entire tab completely die while trying to load :(

**Comment by Bl4Cc4t** [comment]
Sadly I cannot reproduce any of these issues...

Tested on
- Brave 1.42.88
- Chrome 104.0.5112.81
- Firefox 88.0.1 & 103.0.1
- Opera 89.0.4447.83

Are there perhaps any errors showing up in the console?
Please press F12 and then switch to "console". The yellow messages are warnings, they are not related to this script. You can toggle them off by removing the tick under "Custom levels" (Chrome) 
![Inked183266019-c424653c-a6d3-4479-aa05-f680aae24f57_LI](https://user-images.githubusercontent.com/6740726/183266146-3fde26c0-820c-4134-a169-a156c324805b.jpg)

or by clicking "Warnings" so it isn't marked anymore 
![image](https://user-images.githubusercontent.com/6740726/183266078-1c34cfe4-84fe-4a3e-bdcb-5178825bd969.png)

A screenshot of the output there usually helps more than a log.


---

*Offtopic:*
These kinds of bugs are by far the nastiest. The ones others got on their end but cannot be reproduced by me (so I cannot fix them). 
I always suspect twitter to change things on the site to a specific group of accounts. For example, I still have no access to anything related to "communities", a change added a few months ago now. I can't even know if the button I added for that stuff is working now which is just hideous.
Sorry for the rambling.

**Comment by Mavoy** [comment]
Chrome/Opera: "DevTools failed to load source map: Could not load content for https://ton.local.twitter.com/responsive-web-internal/sourcemaps/client-serviceworker/serviceworker.18da1e68.js.map: Connection error: net::ERR_NAME_NOT_RESOLVED"

freezing, not loading further

Firefox:
"Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at https://api.twitter.com/1.1/jot/client_event.json. (Reason: CORS request did not succeed). Status code: (null)."

that's all I'm getting

**Comment by EdinUser** [comment]
Firefox 103.0.1 Ubuntu 22, canonical
![image](https://user-images.githubusercontent.com/5362622/183275316-4486fe6d-fb3c-48f6-971f-f3ceb65f4d98.png)
![image](https://user-images.githubusercontent.com/5362622/183275340-26f3dca9-adf7-4a65-8c59-251ae51ac310.png)

Chromium 104.0.5112.79 (Official Build) snap (64-bit) Ubuntu 22
![image](https://user-images.githubusercontent.com/5362622/183275363-de14a53c-ee52-452e-95db-f98ac4fd66c6.png)
![image](https://user-images.githubusercontent.com/5362622/183275380-846eb3b8-c826-494c-9192-f5bcff99065a.png)

In both of the cases either browser tab crashes, or the browser itself dies.

**Comment by Aridow** [comment]
![image](https://user-images.githubusercontent.com/65539338/183286974-f5e69190-84bd-4847-9b8b-fa5f05975f3a.png)
In English it means:
Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at api.twitter.com... Reason: CORS request did not succeed

**Comment by Bl4Cc4t** [comment]
As I suspected, no errors related to the script...
It looks like the freezing occurs at random during the initial loading of the page, judging by the screenshots.
I'm really sorry, but I certainly don't know how to fix this issue myself...

If someone wants to try and test this debug script I made, that would be helpful: [gt2eb.0.0.40-debug.user.js](https://gist.github.com/Bl4Cc4t/25e49cb4cfee5da8a318d67446f553e3/raw/0659d37c7c6d067482048071b038e862b257c723/gt2eb.0.0.40-debug.user.js)
It just prints the current line to the console for each scoped block. Maybe this helps to pinpoint the bug a bit...
If you try it out, please take a screenshot of the console output and post it here.

**Comment by Aridow** [comment]
Here what it shows for me:
![image](https://user-images.githubusercontent.com/65539338/183295794-8cb9e990-cdf6-4233-a63f-7fd5aced459a.png)
![image](https://user-images.githubusercontent.com/65539338/183295819-1c5aee33-38a9-4a92-870b-a0f0967b0e24.png)
![image](https://user-images.githubusercontent.com/65539338/183295854-e4fcb43b-03a6-4da6-a4dc-f811ea0f6090.png)
![image](https://user-images.githubusercontent.com/65539338/183295880-579aa280-6941-4f73-b48b-38f5feba118b.png)
![image](https://user-images.githubusercontent.com/65539338/183295957-301c3b4b-36ec-4e4f-afd2-42d0ac089609.png)

**Comment by EdinUser** [comment]
Here's mine - the marked lines are where the previous page ends

<details>
<summary>Firefox 103.0.1, Ubuntu 22 console</summary>
![image](https://user-images.githubusercontent.com/5362622/183302205-39695d91-28c9-4cb1-8b8b-18e235217659.png)
![image](https://user-images.githubusercontent.com/5362622/183302215-775caf6f-d646-49de-8ef5-2cd42b948456.png)
![image](https://user-images.githubusercontent.com/5362622/183302231-8f4ee7c5-70fe-4674-aa82-d01e4b8c890a.png)
![image](https://user-images.githubusercontent.com/5362622/183302244-63730ee1-197d-4920-88ab-3c561aed123b.png)
![image](https://user-images.githubusercontent.com/5362622/183302253-45dbd578-d80c-4476-8ff8-c4ff9a87a82d.png)
![image](https://user-images.githubusercontent.com/5362622/183302263-51fcb6d0-b04f-4337-88ec-a80cc8ce74eb.png)
![image](https://user-images.githubusercontent.com/5362622/183302277-537fd6cf-4e53-4bcd-8afe-1352a847efbd.png)
</details>

**Comment by Mavoy** [comment]
the weird part is - sometimes it's working. Like randomly, I went to Twitter, GT2 works just fine, I scrolled the feed, I retweeted some stuff, I switched to another account, used it normally, and then returned to the main account - and bam, it freezes. Shame I didn't think about opening console. I have no idea why it's happening.

**Comment by chillycontro** [comment]
same problem here, goodtwitter 2 was working earlier in the day, but now it's a 50/50 gamble on whether it works fine or is absolutely broken:
<details>
<summary>console view, on firefox v103.0.2 (64-bit), windows 10 home </summary>

![image](https://user-images.githubusercontent.com/28816808/183985804-ac92c33d-26d6-43b9-be30-819471159c87.png)

</details>

**Comment by thatsprettygood** [comment]
I was wondering why my twitter page started hanging..
This happened right after I updated Firefox.
The debug codes on which the script crashes are inconsistent, which coupled with the browser update, indicates that this problem may not be able to be fixed with a script update.

**Comment by Mavoy** [comment]
Honestly, I fear this is the end. It was always coming. I really hate the new layout and I'm depressed I'll have to use it. It's awful of Twitter to not give users choice and force anyone to use a mobile layout on a desktop. Like a mobile website doesn't exist already. It looks abhorrent.

**Comment by EdinUser** [comment]
![nogod](https://user-images.githubusercontent.com/5362622/184379230-b99973f6-aa52-4207-94be-24fd4ad5ee29.gif)

C'mon, somehow it can be made to work again...

**Comment by Bl4Cc4t** [comment]
I am very sure I can fix this issue, but I just cannot reproduce it. Has someone in here with tested if the issues happen with older browser versions?

**Comment by highesttt** [comment]
Whoops, did not see this opened. Chrome 90.0.4430.72
![image](https://user-images.githubusercontent.com/91629626/184420527-cb412b1a-f5e4-4460-b91b-81c4a88698a5.png)

It doesn't even change the look and doesn't load

**Comment by clausies** [comment]
I just downgraded and tested it on Firefox 102.0.1 (64-bit) and experienced the same issue, unfortunately. Since the issue seems to be a bit sporadic, I happen to agree with the consensus that it's userbased and a Twitter side issue.
Not sure if it means anything, but I did try a different Twitter reskin (Lume) which does work, so I don't think Twitter will actively block ALL changes made to their site (although Lume is an extension and the coding is obviously completely different in general as well). Let's not give up hope! Perhaps someone has a side account with the issue that Bl4Cc4t can use to try and reproduce the error? Let's not give up hope!
![102.0.1](https://user-images.githubusercontent.com/50381214/184421967-c73cf830-9606-4ceb-a743-b6170e8f4907.png)

**Comment by highesttt** [comment]
What I find the most weird is that It happened at different times for some people... 
For some of you it has happened for a week, for me its only been a day and nothing was updated

**Comment by Aridow** [comment]
I don't know if it can be useful, but I tried to install [version 0.0.6](https://github.com/Bl4Cc4t/GoodTwitter2/blob/187fd8fd991eab489b9c912f36b0d85093257a59/twitter.gt2eb.user.js) (because it's shorter) of Good Twitter 2 and to play with it (I have no idea how it works). There's this part in it:
```
  let navHome = `nav > a[href='/home'],
                 nav > a[href='/notifications'],
                 nav > a[href='/messages']`
```
and when I replace it by:
```
  let navHome = `nav > a[href='/messages']`
```
...it *seems* to work. It doesn't crash or load endlessly anymore.

I then tried with [version 0.0.21](https://github.com/Bl4Cc4t/GoodTwitter2/blob/f7ca367f8d608b3070afc3546d6d5f8900237019/twitter.gt2eb.user.js) (the last version where this part appears) and it works too. 

Version 0.0.40 is too long to try and delete each part like that, but I randomly deleted all the " // get account information" part anyway, and it works. It breaks, but it loads without crashing. 

I have Firefox 103.0.2 and Windows 11. Hope it can help.

**Comment by EdinUser** [comment]
When I removed the avatar Regex, it shows the feed, nothing else, and the console goes nuts with error for missing avatarUrl:
![image](https://user-images.githubusercontent.com/5362622/184471257-5869e2a2-2d4c-482b-85ca-069df6b34f42.png)

```js
// get account information
  function getInfo() {;GM_log(110)
    let sel = "#react-root ~ script"
    let infoScript = $(sel).text()
    function x(reg, defaultVal="") {;GM_log(113)
      let m = infoScript.match(reg)
      return m ? m[1] : defaultVal
    }
    return {
      bannerUrl:  x(/profile_banner_url\":\"(.+?)\",/),
      avatarUrl:  x(),
      screenName: x(/screen_name\":\"(.+?)\",/, "youarenotloggedin"),
      name:       x(/(?:true|false),\"name\":\"(.+?)\",/, x(/screen_name\":\"(.+?)\",/, "Anonymous")),
      id:         x(/id_str\":\"(\d+)\"/, "0"),
      stats: {
        tweets:    parseInt(x(/statuses_count\":(\d+),/, "0")),
        followers: parseInt(x(/\"followers_count\":(\d+),/, "0")),
        following: parseInt(x(/friends_count\":(\d+),/, "0")),
      }
    }
  }`

Is it possible for some kind of endless loop to appear and this blocks the browser?

**Comment by EdinUser** [comment]
OK, a little bit narrowing down. Somehow I've decided the problem might be in the "navbar". So I played a little with it, first disabling all calls to the function `addNavBar()`, and later - commenting only `for` loop in it. This way Twitter loads with a little broken navbar on top, GT accepts changing of settings, etc. I will try to check which part is exactly the culprit of the loop...

```js
  // add navbar
  function addNavbar() {
    waitForKeyElements(`nav > a[href="/home"]`, () => {
      if ($(".gt2-nav").length) return

      $("main").before(`
        <nav class="gt2-nav">
          <div class="gt2-nav-left"></div>
          <div class="gt2-nav-center">
            <a href="/home"></a>
          </div>
          <div class="gt2-nav-right">
            <div class="gt2-search"></div>
            <div class="gt2-toggle-navbar-dropdown">
              <img src="${getInfo().avatarUrl.replace(/_(bigger|normal|(reasonably_)?small|\d*x\d+)/, "_bigger")}" />
            </div>
            <div class="gt2-compose">${getLocStr("composeNewTweet")}</div>
          </div>
        </nav>
        <div class="gt2-search-overflow-hider"></div>
      `)

      // home, notifications, messages
        /*
      for (let e of [
        "Home",
        "Notifications",
        "Messages",
        window.innerWidth < 1005 ? "Explore" : null
      ]) {
        if (!e) continue
        let $e = $(`nav > a[href^="/${e.toLowerCase()}"]:not([data-testid=AppTabBar_Profile_Link]):not([href$="/lists"])`)
        if (!e.length) continue
        $e.appendTo(".gt2-nav-left")
        $(`.gt2-nav a[href^="/${e.toLowerCase()}"] > div`)
        .append(`
          <div class="gt2-nav-header">
            ${getLocStr(`nav${e}`)}
          </div>
        `)
        .attr("data-gt2-color-override-ignore", "")
      }
*/
      // highlight current location
      $(`.gt2-nav a[href^='/${getPath().split("/")[0]}']`).addClass("active")

      // twitter logo
      $("h1 a[href='/home'] svg")
      .appendTo(".gt2-nav-center a")
    })
  }
```

**Comment by Bl4Cc4t** [comment]
Thank you @Aridow @EdinUser so much for finding the source of this error!
Since this seems to work for both of you, I'm guessing it really is the culprit.
I'll put an update out ASAP since this issue affects so many people.


I am really sorry for not being able to fix this myself, but somehow my accounts are on some less updated site all the time - not sure what to do about it.

I plan on removing jQuery altogether and restructuring the project a bit, it should (hopefully) become easier in the future to find these kinds of errors!

**Comment by Aridow** [comment]
I confirm @EdinUser's fix works for me.

**Comment by JesperDramsch** [comment]
Can confirm as well. The change of line 806 does it for me. Thank you @EdinUser!

**Comment by Bl4Cc4t** [comment]
Awesome, will release an update in a minute!

**Comment by Mavoy** [comment]
Indeed, it's also working for me - I mean the timeline is working normally, but the navigation bar disappeared? Could be because I'm on 0.0.40?

Sorry for depression strike yesterday, should've known better, especially since I'm spamming you guys with every reply

**Comment by Bl4Cc4t** [comment]
just released the update. 

@Mavoy please let me know if you have any issues with v0.0.40.1 as well!

**Comment by jeffyjeffy1023** [comment]
To anyone else that was about to make a new issue due to this bug seemingly coming back: Try just doing a hard refresh (CTRL+SHIFT+R, or COMMAND+SHIFT+R on a Mac) and that should fix it. Fixed it for me at least here on Firefox 114.0.2 with GT2 0.0.44.1.
